Add linters configuration, reformat whole code

antoni-szych-rtbhouse opened the pull request:
Hi,

As black and isort became the standard tools to enforce uniform style within the whole codebase, I've came to a conclusion that these may be helpful also here, as the style currently varies in different parts of the codebase.

I've added:
- black
- isort
- moved the pytest config to pyproject.toml

I've reformated the code according to black and isort configuration.

Topics that are very open for discussion:
- line lengths. Black gives default of 88 characters per line. ~Given that nearly all displays currentlyt have 16:9 / 16:10 aspect ratio instead of 4:3, 120 characters per line also seems reasonable (which I've set in the configuration in pyproject.toml)~ **Default of 88 chars is used**. However this setting is entirely dependent on the personal preference and can be easily changed to any other value. [more reading](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length)
- string normalization. Black by default converts all strings that do not contain double quotes `"` to double-quoted ones instead of singe-quoted `'`. However, some people have strong opinion regarding quoting style and allow double-quotes only in docstrings. ~I've left the default normalization on, but performed the normalization in separate commit. If needed, that commit can be easily skipped and the configuration can be changed to include `skip-string-normalization`~. [more reading](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#strings)
String normalization was dropped from this PR.

ds-cbo commented:
"I've left the default normalization on"

I don't have strong opinions on quotes, but I do have a strong opinion on git history: Please try to keep the changes minimal when introducing a new tool. Other developers will `git annotate` a file and 50% of the lines will get "black 2023/12/14" as annotation, which is completely useless.

ds-cbo commented on voluptuous/__init__.py:
```diff
@@ -1,9 +1,9 @@
-from voluptuous.validators import *
 from voluptuous.util import *
-from voluptuous.error import *
+from voluptuous.validators import *
```
This might actually be a breaking change if anything in `.error` depends on anything in `.validators` (which is likely), I would disable isort for `__init__.py` files

antoni-szych-rtbhouse replied:
```
from voluptuous.error import * # isort: skip
```
made `error` go to the bottom of imports

ds-cbo commented on voluptuous/schema_builder.py:
```diff
@@ -148,11 +150,24 @@ def Extra(_) -> None:
+    object,
+    dict,
+    None,
+    typing.Callable,
```
this had intentional compact formatting (similar types on the same line), please wrap it in `# fmt: off` and `# fmt: on`

alecthomas replied:
Agreed.

antoni-szych-rtbhouse replied:
sure thing

ds-cbo commented on voluptuous/schema_builder.py:
```diff
@@ -854,11 +869,13 @@ def is_callable(key_):
+        (4, is_type),  # types/classes lowest before Extra
+        (3, is_callable),  # callables after markers
+        (5, is_extra),
+    ]  # Extra lowest priority
```
move this comment back to the respective line

antoni-szych-rtbhouse replied:
sure thing.

ds-cbo commented on voluptuous/schema_builder.py:
```diff
@@ -1237,9 +1274,7 @@ def _args_to_dict(func, args):
-    arguments = dict((arg_name, arg_value_list[i])
-                     for i, arg_name in enumerate(arg_names)
-                     if i < len(arg_value_list))
+    arguments = dict((arg_name, arg_value_list[i]) for i, arg_name in enumerate(arg_names) if i < len(arg_value_list))
```
I wouldn't say that this is more readable now, please wrap it back to several lines (decreasing line length as per my other suggestion would probably help already)

antoni-szych-rtbhouse replied:
limiting the line length to default 88 did the trick

ds-cbo commented on voluptuous/tests/tests.py:
```diff
@@ -1,26 +1,77 @@
+    UrlInvalid,
+    raises,
+    validate,
+)
```
please compact this down as well, with `# fmt: off` and `# fmt: on` to keep it in place
no one likes a 50-line import statement

antoni-szych-rtbhouse replied:
sure thing

ds-cbo commented on voluptuous/tests/tests.py:
```diff
@@ -1300,39 +1319,36 @@ def __hash__(self):
-                                          extra=ALLOW_EXTRA))},
-                    extra=ALLOW_EXTRA)
+    schema = Schema(
+        {"number": int, "follow": All(Self, Schema({"extra_number": int}, extra=ALLOW_EXTRA))}, extra=ALLOW_EXTRA
```
this also didn't get any clearer, please expand it again (might be able to get black to do that with a trailing comma)

antoni-szych-rtbhouse replied:
limiting the line length to default 88 did the trick

ds-cbo commented on voluptuous/validators.py:
```diff
@@ -1,18 +1,41 @@
+    TrueInvalid,
+    TypeInvalid,
+    UrlInvalid,
+)
```
please compact this one down as well, 20-line imports are still no fun

antoni-szych-rtbhouse replied:
yep, there is, `multi_line_output = 5` - however it's incompatible with black if it gets longer than `line_length`. `# fmt: off` and `# fmt: on` to exclude it from black formatting gets the job done.

alecthomas commented:
> Counterpoint: I often have 2 terminals side-to-side on my 16:9 screen, and then it only fits up to 100 characters comfortably. It also depends on font-size, which is still very personal nowadays. I would keep the default value of black (80 + 10% wrap margin)

Counter-counterpoint: I suspect the majority of people, myself included, prefer 100 or 120 columns, and for the minority who don't, having to scroll horizontally every now and then is not the end of the world. I have to do it myself sometimes on smaller screens, and I'm fine with it.

ds-cbo commented on voluptuous/schema_builder.py:
```diff
@@ -937,8 +954,7 @@ class Msg(object):
         if cls and not issubclass(cls, er.Invalid):
-            raise er.SchemaError("Msg can only use subclases of"
-                                 " Invalid as custom class")
+            raise er.SchemaError("Msg can only use subclases of" " Invalid as custom class")
```
I've had this before, the thing with black is that it 100% guarantees that the AST stays the exact same between transformations, and merging two strings would break that promise on python 3.7- (see https://github.com/psf/black/issues/26 )

But I think it should detect that we're not running on python 3.7- and merge them, so I'm not sure what's going on there

antoni-szych-rtbhouse replied:
You're right, it does not out of the box [(yet, but it should starting with black 24)](https://github.com/psf/black/issues/1837). I will run it with the mentioned experimental flag and manually check output before commiting.

[flake8-implicit-str-concat](https://pypi.org/project/flake8-implicit-str-concat/) would be useful for such cases, I can add it to the flake8 tox env if you agree.
I'm using pylint on a daily basis, it also [has this rule](https://pylint.readthedocs.io/en/stable/user_guide/messages/warning/implicit-str-concat.html)

antoni-szych-rtbhouse commented on voluptuous/schema_builder.py:
```diff
@@ -296,24 +313,22 @@ def _compile(self, schema):
+            key
+            for key in schema
+            if key is not Extra
+            and ((self.required and not isinstance(key, (Optional, Remove))) or isinstance(key, Required))
```
I'm afraid it's not configurable. Black has (by design and purpose) very limited configuration options, boiling down to:
- line length
- string quotes normalization on or off
- "magic trailing comma" on or off

The upside is that you don't have to choose and maintain your own formatting style in a project.

antoni-szych-rtbhouse commented:
> "I've left the default normalization on"
> 
> I don't have strong opinions on quotes, but I do have a strong opinion on git history: Please try to keep the changes minimal when introducing a new tool. Other developers will `git annotate` a file and 50% of the lines will get "black 2023/12/14" as annotation, which is completely useless.

This can be actually [easily fixed both locally and for github blame](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html#avoiding-ruining-git-blame): we just need to put the commit sha into the `.git-blame-ignore-revs` file. Note that it will have to be added after merging this PR, as we don't know the merge commit sha in advance :)

antoni-szych-rtbhouse commented on voluptuous/validators.py:
```diff
@@ -33,42 +56,39 @@
     # domain
-    r'(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+'
-    r'(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?$)'
+    r"(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+" r"(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?$)"
```
sure. Note that this regex is long obsolete due to tlds like `.basketball`, `.google` or `.youtube`:
- https://en.wikipedia.org/wiki/List_of_Internet_top-level_domains#ICANN-era_generic_top-level_domains
- https://en.wikipedia.org/wiki/List_of_Internet_top-level_domains#Brand_top-level_domains

antoni-szych-rtbhouse commented:
Hi @ds-cbo @alecthomas can you please take a look at the updated version? :)

ds-cbo commented:
I have no new complaints, but I'm also not responsible for this project

antoni-szych-rtbhouse commented:
@alecthomas could you please take a look? :) I think all the major obstacles were resolved with the change in line length and few `fmt:off/on`s and `ignore`s.

alecthomas commented:
Mmm yeah, one last change please: can you make the default string quoting `'` instead of `"`. That is the default in this project, and leaving it as-is will reduce the diff size considerably.

alecthomas commented on voluptuous/validators.py:
```diff
@@ -33,42 +40,42 @@
     # dot-atom
-    r"(^[-!#$%&'*+/=?^_`{}|~0-9A-Z]+"
-    r"(\.[-!#$%&'*+/=?^_`{}|~0-9A-Z]+)*$"
+    r"(^[-!#$%&'*+/=?^_`{}|~0-9A-Z]+(\.[-!#$%&'*+/=?^_`{}|~0-9A-Z]+)*$"
```
Why did this get joined? Can we split this again please?

antoni-szych-rtbhouse replied:
because these 2 lines can fit into one line, black merged them into:
```
    r"(^[-!#$%&'*+/=?^_`{}|~0-9A-Z]+" r"(\.[-!#$%&'*+/=?^_`{}|~0-9A-Z]+)*$"
```
which raises obvious issue "why the strings were not concatenated, which results in:
```
    r"(^[-!#$%&'*+/=?^_`{}|~0-9A-Z]+(\.[-!#$%&'*+/=?^_`{}|~0-9A-Z]+)*$"
```
The only way to keep the original version is `# fmt: on` and `# fmt: off`. I've added it, so these lines are back to being separate.

alecthomas commented on voluptuous/validators.py:
```diff
@@ -33,42 +40,42 @@
     # literal form, ipv4 address (SMTP 4.1.3)
-    r'|^\[(25[0-5]|2[0-4]\d|[0-1]?\d?\d)'
-    r'(\.(25[0-5]|2[0-4]\d|[0-1]?\d?\d)){3}\]$'
+    r'|^\[(25[0-5]|2[0-4]\d|[0-1]?\d?\d)(\.(25[0-5]|2[0-4]\d|[0-1]?\d?\d)){3}\]$'
```
Same thing here.

antoni-szych-rtbhouse replied:
I've added `# fmt: on` and `# fmt: off`, so these lines are back to being separate.

antoni-szych-rtbhouse commented:
> Mmm yeah, one last change please: can you make the default string quoting `'` instead of `"`. That is the default in this project, and leaving it as-is will reduce the diff size considerably.

Sure, I've removed the string normalization commit (it was the last one) and the PR is smaller (`+755 −433` vs `+1,047 −730` previously).

I've also made a test and replaced all possible double quotes to single quotes in 231 lines. This would cause the diff to be `+966 −648`. Please let me know if case you would like me to commit this.

alecthomas reviewed (approved):
Thanks for doing this!

ds-cbo commented:
> This can be actually [easily fixed both locally and for github blame](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html#avoiding-ruining-git-blame): we just need to put the commit sha into the .git-blame-ignore-revs file. Note that it will have to be added after merging this PR, as we don't know the merge commit sha in advance :)

Now please don't forget to do that as well, now we have this:

<img width="475" alt="image" src="https://github.com/alecthomas/voluptuous/assets/82801887/0db08f95-e0a9-4ec1-93ae-8605aa02ed10">

antoni-szych-rtbhouse commented:
> Now please don't forget to do that as well, now we have this:

Done in #505, it will look like following (screenshot from [my fork with this commit already merged](https://github.com/antoni-szych-rtbhouse/voluptuous/blame/master/voluptuous/humanize.py)):

![image](https://github.com/alecthomas/voluptuous/assets/84847001/2e5f5321-4eef-4b71-b4f7-4cafb193d63c)


As you can see, there are some lines annotated with the refactor commit, however these are only the new lines, which did not exist at all previously. However the important lines like 26-31, 36-40, 45-46, 51-53 show the pre-refactor annotation.
